cmd/zrepl/zrepl.c

```diff
@@ -190,27 +249,25 @@ uzfs_zvol_io_receiver(void *arg)
 	}
 exit:
 	if (zinfo != NULL) {
-		ZREPL_LOG("uzfs_zvol_io_receiver thread exiting, volume:%s\n",
+		ZREPL_LOG("uzfs_zvol_io_receiver thread exiting, volume: %s\n",
 		    zinfo->name);
 		(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
 		zinfo->conn_closed = B_TRUE;
 		/*
 		 * Send signal to ack sender so that it can free
 		 * zio_cmd, close fd and exit.
 		 */
-		(void) pthread_mutex_lock(&zinfo->complete_queue_mutex);
 		if (zinfo->io_ack_waiting) {
 			rc = pthread_cond_signal(&zinfo->io_ack_cond);
 		}
-		(void) pthread_mutex_unlock(&zinfo->complete_queue_mutex);
+		(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
 		/*
```

Comment:
I will slightly change the code here to hold mutex when ack_sender_created flag is checked, but I'm not sure I understand your concern here. Which comments do you mean?

Reply:
This is the comment I was referring to:
                /*
                 * wait for ack thread to exit to avoid races with new
                 * connections for the same zinfo
                 */

Reply:
The basic assumption for correct function must be that whenever `zinfo->is_io_ack_sender_created` and `zinfo->conn_closed` are read or set, the zinfo mutex must be held. So this is what I was trying to fix. In particular I observed problem with following scenario: zvol_io_receiver is started while previous zvol_io_receiver is exiting (and previous ack_sender is running). Unlikely that all possible race conditions for this case are fixed but I think at least the most obvious are fixed.